packages/jest-cli/src/runJest.js

```diff
@@ -9,108 +9,161 @@
+    ? chalk.dim(' in ') + contexts.length + chalk.dim(' projects')
     : '';
 
   return chalk.dim('Ran all test suites') +
     testInfo +
     nameInfo +
+    contextInfo +
     chalk.dim('.');
 };
 
+const getNoTestsFoundMessage = (testRunData, pattern) => {
+  if (pattern.onlyChanged) {
+    return chalk.bold(
+      'No tests found related to files changed since last commit.\n',
+    ) +
+      chalk.dim(
+        pattern.watch
+          ? 'Press `a` to run all tests, or run Jest with `--watchAll`.'
+          : 'Run Jest without `-o` to run all tests.',
+      );
+  }
+
+  const pluralize = (word: string, count: number, ending: string) =>
+    `${count} ${word}${count === 1 ? '' : ending}`;
+  const testPathPattern = formatTestPathPattern(pattern);
+  const individualResults = testRunData.map(testRun => {
+    const stats = testRun.matches.stats || {};
+    const config = testRun.context.config;
+    const statsMessage = Object.keys(stats)
+      .map(key => {
+        if (key === 'roots' && config.roots.length === 1) {
+          return null;
+        }
+        const value = config[key];
+        if (value) {
+          const matches = pluralize('match', stats[key], 'es');
+          return `  ${key}: ${chalk.yellow(value)} - ${matches}`;
+        }
+        return null;
+      })
+      .filter(line => line)
+      .join('\n');
+
+    return testRun.matches.total
+      ? `In ${chalk.bold(config.rootDir)}\n` +
+          `  ${pluralize('file', testRun.matches.total || 0, 's')} checked.\n` +
+          statsMessage
+      : `No files found in ${config.rootDir}.\n` +
+          `Make sure Jest's configuration does not exclude this directory.` +
+          `\nTo set up Jest, make sure a package.json file exists.\n` +
+          `Jest Documentation: ` +
+          `facebook.github.io/jest/docs/configuration.html`;
+  });
+  return chalk.bold('No tests found') +
+    '\n' +
+    individualResults.join('\n') +
+    '\n' +
+    `Pattern: ${chalk.yellow(testPathPattern)} - 0 matches`;
+};
+
+const getTestPaths = async (context, pattern, argv, pipe) => {
+  const source = new SearchSource(context);
+  let data = await source.getTestPaths(pattern);
+  if (!data.paths.length) {
+    const localConsole = new Console(pipe, pipe);
+    if (pattern.onlyChanged && data.noSCM) {
+      if (context.config.watch) {
+        // Run all the tests
+        setState(argv, 'watchAll', {
+          noSCM: true,
+        });
+        pattern = getTestPathPattern(argv);
+        data = await source.getTestPaths(pattern);
+      } else {
+        localConsole.log(
+          'Jest can only find uncommitted changed files in a git or hg ' +
+            'repository. If you make your project a git or hg ' +
+            'repository (`git init` or `hg init`), Jest will be able ' +
+            'to only run tests related to files changed since the last ' +
+            'commit.',
+        );
+      }
+    }
+  }
+
+  return data;
+};
+
 const runJest = async (
```

Comment:
can you rewrite this function as something like this and take everything else out?
```js
runJest = async (/* ... */) => {
  const maxWorkers = getMaxWorkers(argv);
  const testPathPattern = getTestPathPattern(argv);
  const firstContext = contexts[0];
  const testRunData = await getTestRunData(/*...*/);
  const sortedTestsFromAllProjects = getSortedTestsFromAllProjects(/*...*/);
  warnIfNoTestsFound(/*...*/);
  setVerboseIfRunningOnlyOneTest(contexts);
  const testResults = await runTests(/*...*/);
  return processResults(testResults);
};
```

sorry :( very hard to review and read the code that has dozens of branches :(